include/oneapi/dpl/internal/distributed_ranges_impl/shp/algorithms/sort.hpp

```diff
@@ -137,118 +139,125 @@ sort(R&& r, Compare comp = Compare())
      }).wait();
 
     std::vector<std::size_t*> splitter_indices;
-    std::vector<std::size_t> sorted_seg_sizes(n_splitters + 1);
+    // sorted_seg_sizes[i]: how many elements exists in all segments between
+    // medians[i-1] and medians[i]
+    std::vector<std::size_t> sorted_seg_sizes(n_segments, 0);
+    // push_positions[snd_idx][rcv_idx]: shift inside final segment of rcv_idx for
+    // data being sent from initial snd_idx segment
     std::vector<std::vector<std::size_t>> push_positions(n_segments);
 
     // Compute how many elements will be sent to each of the new "sorted
     // segments". Simultaneously compute the offsets `push_positions` where each
     // segments' corresponding elements will be pushed.
 
-    segment_id = 0;
-    for (auto&& segment : segments)
+    for (auto&& [segment_id, segment] : rng::views::enumerate(segments))
     {
         auto&& q = __detail::queue(ranges::rank(segment));
         auto&& local_policy = __detail::dpl_policy(ranges::rank(segment));
 
         auto&& local_segment = __detail::local(segment);
 
-        std::size_t* splitter_i = sycl::malloc_shared<std::size_t>(n_splitters, q.get_device(), context());
+        // splitter_i = [ index in local_segment of first element greater or equal
+        // 1st global median, index ... 2nd global median, ..., size of
+        // local_segment]
+        std::size_t* splitter_i = sycl::malloc_shared<std::size_t>(n_segments, q.get_device(), shp::context());
         splitter_indices.push_back(splitter_i);
 
-        // Local copy `medians_l` necessary due to bug in Level Zero.
-        T* medians_l = sycl::malloc_device<T>(n_splitters, q.get_device(), context());
+        // Local copy `medians_l` necessary due to [GSD-3893]
```

Comment:
We can not use jira ID in the oneDPL code. Please restore previous version of this line.